atc/metric/emitter/newrelic.go

```diff
@@ -173,16 +144,117 @@ func (emitter *NewRelicEmitter) Emit(logger lager.Logger, event metric.Event) {
 	// otherwise recording it. (This won't be easily graphable, that's okay,
 	// this is more for monitoring synthetics)
 	if event.State != metric.EventStateOK {
-		singlePayload := emitter.simplePayload(logger, event, "alert")
+		singlePayload := emitter.transformToNewRelicEvent(logger, event, "alert")
 		// We don't have friendly names for all the metrics, and part of the
 		// point of this alert is to catch events we should be logging but
 		// didn't; therefore, be consistently inconsistent and use the
 		// concourse metric names, not our translation layer.
 		singlePayload["metric"] = event.Name
-		payload = append(payload, singlePayload)
+		emitter.NewRelicBatch = append(emitter.NewRelicBatch, singlePayload)
+	}
+
+	duration := time.Since(emitter.LastEmitTime)
+	if len(emitter.NewRelicBatch) >= emitter.BatchSize || duration >= emitter.BatchDuration {
+		logger.Debug("pre-emit-batch", lager.Data{
+			"batch-size":         emitter.BatchSize,
+			"current-batch-size": len(emitter.NewRelicBatch),
+			"batch-duration":     emitter.BatchDuration,
+			"current-duration":   duration,
+		})
+		emitter.submitBatch(logger)
+	}
+}
+
+// NewRelic has strict requirements around the structure of the events
+// Keys must be alphanumeric and can contain hyphens or underscores
+// Values must be sting, int, or unix timestamps. No maps/arrays.
+func (emitter *NewRelicEmitter) transformToNewRelicEvent(logger lager.Logger, event metric.Event, nameOverride string) NewRelicEvent {
+	name := nameOverride
+	if name == "" {
+		name = strings.Replace(event.Name, " ", "_", -1)
+	}
+
+	eventType := fmt.Sprintf("%s%s", emitter.prefix, name)
+
+	payload := NewRelicEvent{
+		"eventType": eventType,
+		"value":     event.Value,
+		"state":     string(event.State),
+		"host":      event.Host,
+		"timestamp": event.Time.Unix(),
+	}
+
+	for k, v := range event.Attributes {
+		payload[fmt.Sprintf("_%s", k)] = v
 	}
+	return payload
+}
+
+func (emitter *NewRelicEmitter) submitBatch(logger lager.Logger) {
+	batchToSubmit := make([]NewRelicEvent, len(emitter.NewRelicBatch))
+	copy(batchToSubmit, emitter.NewRelicBatch)
+	emitter.NewRelicBatch = make([]NewRelicEvent, 0)
+	emitter.LastEmitTime = time.Now()
+	go emitter.emitBatch(logger, batchToSubmit)
+}
+
+func (emitter *NewRelicEmitter) emitBatch(logger lager.Logger, payload []NewRelicEvent) {
+	batch, err := emitter.marshalJSON(logger, payload)
+	if err != nil {
+		logger.Error("failed-to-marshal-batch", err)
+		return
+	}
+
+	req, err := http.NewRequest("POST", emitter.Url, batch)
+	if err != nil {
+		logger.Error("failed-to-construct-request", err)
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("X-Insert-Key", emitter.apikey)
+	if !emitter.CompressionDisabled {
+		req.Header.Add("Content-Encoding", "gzip")
+	}
+
+	resp, err := emitter.Client.Do(req)
```

Comment:
I wonder if we should put a timeout on these requests 😬  

might be better in a different issue though - wdyt?

Reply:
@pivotal-bin-ju mentioned that we already set a timeout when we create the `client`. See line 61.